### ISSUE TITLE
False positive with ns_number_init_as_function_reference

### ISSUE THREAD

**Issue by khsora34** [realm/SwiftLint#4574]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

When I try to use the nil coalescing operator with NSNumber initializer, the rule ns_number_init_as_function_reference is triggered.
I´m using the `NSNumber.init(value:)` initializer, as asked by the rule. 

I was just updating SwiftLint and this new case appeared in my code. I tried to reproduce it in a file with only that code and it seems to be still triggered. 

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint Example.swift 
Linting Swift files at paths Example.swift
Linting 'Example.swift' (1/1)
/Users/setoncios/Desktop/Example.swift:8:34: warning: NSNumber Init as Function Reference Violation: Passing `NSNumber.init` or `NSDecimalNumber.init` as a function reference is dangerous as it can cause the wrong initializer to be used, causing crashes. Use `.init(value:)` instead. (ns_number_init_as_function_reference)
/Users/setoncios/Desktop/Example.swift:9:34: warning: NSNumber Init as Function Reference Violation: Passing `NSNumber.init` or `NSDecimalNumber.init` as a function reference is dangerous as it can cause the wrong initializer to be used, causing crashes. Use `.init(value:)` instead. (ns_number_init_as_function_reference)
/Users/setoncios/Desktop/Example.swift:6:1: warning: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
Done linting! Found 3 violations, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? **0.50.0**
* Installation method used (Homebrew, CocoaPods, building from source, etc)? **Homebrew**
* Paste your configuration file:

```yml
# no configuration file, using default configuration
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? **No**
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcodebuild -version`)? **Xcode 14.1 - Build version 14B47b**
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
import Foundation

class Example {
    private let first: NSNumber
    private let second: NSNumber
    
    init(numberA: NSNumber?, numberB: NSNumber?) {
        // This triggers a violation:
        self.first = numbersA ?? NSNumber.init(value: 0)
        // This triggers a violation:
        self.second = numberB ?? NSNumber.init(value: 0)
    }
}
```
